### PR TITLE
Online Prediction Template

### PR DESCRIPTION
**Pull request by svangapally-cs**
https://cognitivescale.atlassian.net/browse/FAB-1847

### CODE REVIEW

File: online-prediction/Makefile [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+tests: check-env
+	cortex skills invoke --params-file ./test/payload.json ${SKILL_NAME} request --project ${PROJECT_NAME}
+
+all: build push deploy
+
+get: check-env
+	cortex skills describe ${SKILL_NAME} --project ${PROJECT_NAME}
+
+check-env:
+ifndef DOCKER_PREGISTRY_URL
```
Comment (yreddy-CS) [inline online-prediction/Makefile:25]: We can default the registry if not explicitly provided to the one that is returned in the info endpoint endpoints.registry

File: online-prediction/predict/main.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    logging.info("Online Prediction: Invoke Request:{}".format(request))
+    columns = request['payload']["columns"]
+    instances = request['payload']["instances"]
+    df = pd.DataFrame(columns=columns, data=instances)
+    # Initialize Cortex Client
+    client = Cortex.client(api_endpoint=request["apiEndpoint"], token=request["token"], project=request["projectId"])
+
+    # Load Model from the experiment run
+    logging.info("Loading model artifacts from experiment run...")
+    model = await load_model(client, request["properties"]["experiment-name"], request["properties"]["run-id"])
```
Comment (yreddy-CS) [inline online-prediction/predict/main.py:31]: Do we need to load the model for every invoke? We can load it once and keep using it

File: online-prediction/predict/main.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    logging.info("Model Loaded!")
+    try:
+        # If the model artifact is of type `dict`
+        if isinstance(model, dict):
+            cat_cols = model["cat_columns"] if "cat_columns" in model else []
+            num_cols = [x for x in df.columns if x not in cat_cols]
```
Comment (yreddy-CS) [inline online-prediction/predict/main.py:37]: better to use variables with proper names

File: online-prediction/predict/main.py (hunk 2 above)
Comment (svangapally-cs) [reply online-prediction/predict/main.py:31]: What if they update the model artifact in the experiment & invoke again?? Is there a way to know that??

File: online-prediction/predict/main.py (hunk 3 above)
Comment (svangapally-cs) [reply online-prediction/predict/main.py:37]: I used the variables to match with german credit train & predict example but yeah I can use the full names here.

File: online-prediction/Makefile (hunk 1 above)
Comment (svangapally-cs) [reply online-prediction/Makefile:25]: Makes sense. I guess we have change this in generate skills templates as well.

Comment (svangapally-cs) [reply online-prediction/Makefile]: Updated Makefile to handle this

**Review by mcriscolo-cs — APPROVED** [review]

File: online-prediction/predict/main.py (hunk 2 above)
Comment (dvasani-CS) [reply online-prediction/predict/main.py:31]: get last run if run_id is not provided, this way we can get an updated artifact. 
we don't have any way to know that artifact is updated.

File: online-prediction/predict/main.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            predictions = model.predict(instances)
+    except Exception as e:
+        raise Exception("Error occurred while making predictions, Please check the model. Message: {}".format(e))
+    return {'payload': predictions.tolist()}
+
+
+async def load_model(client, experiment_name, run_id):
+    try:
+        experiment = client.experiment(experiment_name)
+        run = experiment.get_run(run_id)
```
Comment (dvasani-CS) [inline online-prediction/predict/main.py:62]: check here if run_id is None.. and get Last Run

File: online-prediction/skill.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  - name: experiment-name
+    title: Experiment Name
+    description: Experiment name to retrieve experiment details
+    required: false
+    type: string
+    secure: false
+  - name: run-id
+    title: Run ID
+    description: Run id of the experiment
+    required: true
```
Comment (dvasani-CS) [inline online-prediction/skill.yaml:51]: make run-id optional

**Review by dvasani-CS — CHANGES_REQUESTED** [review]

File: online-prediction/predict/main.py (hunk 2 above)
Comment (svangapally-cs) [reply online-prediction/predict/main.py:31]: @dvasani-CS  The reason being we are using this template for MLops skill wizard flow and we always get a run-id when we create a model skill. Anyway for general purposes, I'm making it optional.

**Review by dvasani-CS — APPROVED** [review]

**Pull request merged**